designs/mixins.md

```diff
@@ -0,0 +1,1075 @@
+{
+    "smithy": "1.1",
+    "shapes": {
+        "smithy.example#MyMixin": {
+            "type": "structure",
+            "members": {
+                "a": {
+                    "target": "smithy.api#String",
+                    "traits": {
+                        "smithy.api#documentation": "A docs"
+                    }
+                },
+                "b": {
+                    "target": "smithy.api#String",
+                    "traits": {
+                        "smithy.api#documentation": "B docs"
+                    }
+                }
+            },
+            "traits": {
+                "smithy.api#mixin": {}
+            }
+        },
+        "smithy.example#MyStruct": {
+            "type": "structure",
+            "mixins": [
+                {"target": "smithy.example#MyMixin"}
+            ],
+            "members": {
+                "a": {
+                    "target": "smithy.api#String",
+                    "traits": {
+                        "smithy.api#documentation": "More specific docs for a",
+                        "smithy.api#override": {}
+                    }
+                },
+                "b": {
+                    "target": "smithy.api#String",
+                    "traits": {
+                        "smithy.api#documentation": "More specific docs for b",
+                        "smithy.api#override": {}
+                    }
+                }
+            }
+        }
+    }
+}
+```
+
+
+### Mixins are not code generated
+
+Mixins are an implementation detail of models and are only intended to reduce
+duplication in Smithy structure and union definitions. Shapes marked with the
+`@mixin` trait MUST NOT be code generated. Mixins _do not_ provide any kind of
+runtime polymorphism for types generated from Smithy models. Generating code
+from mixins removes the ability for modelers to introduce and even remove
+mixins over time as a model is refactored.
+
+
+### Mixins cannot be referenced other than as mixins to other shapes
+
+To ensure that mixins are not code generated, mixins MUST NOT be referenced
+from any other shapes except to mix them into structures and unions. Mixins
+MUST NOT be used as operation input, output, or errors, and they MUST NOT be
+targeted by members.
+
+The following model is invalid because a structure member targets a mixin:
+
+```
+@mixin
+structure GreetingMixin {
+    greeting: String
+}
+
+structure InvalidStructure {
+    notValid: GreetingMixin // <- this is invalid
+}
+```
+
+The following model is invalid because an operation attempts to use a mixin
+as input:
+
+```
+@mixin
+structure InputMixin {}
+
+operation InvalidOperation {
+    input: InputMixin // <- this is invalid
+}
+```
+
+
+### Mixins in the IDL
+
+To support mixins, the `structure_statement` and `union_statement` ABNF rules
+will be updated to contain an optional `with` clause that comes after the
+shape name and before `{`. If present, the `with` clause MUST contain one or
```

Comment:
How do `with` list members scale? Say I had a dozen or so mixins used on a particular shape, would they all be required to be defined in a single line? Or is there a way to breakup the definition like:

```
structure MyStruct with (
    MyMixinA,
    MyMixinB,
    MyMixinC,
...
) {
    ...
}
```

Reply:
In the current proposal, whitespace and commas are insignificant, you can break them up onto different lines, and there's no need for a `(` or `)`:

```
structure MyStruct with
    MyMixinA
    MyMixinB
    MyMixinC
    MyMixinD
{
    @required
    foo: String
}
```

Reply:
Cool, makes sense.